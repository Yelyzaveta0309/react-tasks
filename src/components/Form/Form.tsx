import React, {FC, memo, useState} from 'react';
import { Input } from '@mui/material';
import { Button } from './components/Button';
import { useParams } from 'react-router-dom';
import { getMessageListById } from '../../services/firebase';
import { nanoid } from '@reduxjs/toolkit';
import { push } from 'firebase/database';

export const Form: FC = memo(() => {
    const [value, setValue] = useState('');
    const { chatId } = useParams();
  
    const handleSubmitForm = (e: React.FormEvent<HTMLFormElement>) => {
      e.preventDefault();
  
      if (chatId && value) {
        const id = nanoid();
        push(getMessageListById(chatId), {
          author: 'user',
          id, 
          text: value
        });
      }
      setValue('');
    };
  
    return (
      <form onSubmit={handleSubmitForm}>
        <Input
          type="text"
          value={value}
          onChange={(e) => setValue(e.target.value)}
        />
        <Button disabled={!value} />
      </form>
    );
  });