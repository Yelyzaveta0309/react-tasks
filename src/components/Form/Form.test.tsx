import React from 'react'
import { Form } from './Form';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

describe('Form', () => {

    it('render component', () => {
        const { getByRole } = render(<Form />);
        const inputElement = getByRole('textbox');
        const buttonElement = getByRole('button');
        expect(inputElement).toBeInTheDocument();
        expect(buttonElement).toBeInTheDocument();
    });

    it('render with snapshot', () => {
        const {asFragment} = render(<Form />)
        expect(asFragment()).toMatchSnapshot();
    });

    it('should update input value on change', () => {
        const { getByRole } = render(<Form />);
        const input = getByRole('textbox');
        fireEvent.change(input, { target: { value: 'test' } });
        expect(input).toHaveValue('test');
    });

    it('button is disabled when input value is empty', () => {
        const { getByRole } = render(<Form />);
        const inputElement = getByRole('textbox');
        const buttonElement = getByRole('button');
        expect(buttonElement).toBeDisabled();
        fireEvent.change(inputElement, { target: { value: 'test message' }})
    });

    it('clears the input value when the form is submitted', () => {
        const { getByRole } = render(<Form />);
        const input = getByRole('textbox') as HTMLInputElement;
        const button = getByRole('button') as HTMLButtonElement;
    
        fireEvent.change(input, { target: { value: 'Hello, world!' } });
        fireEvent.click(button);
    
        expect(input.value).toBe('');
    });
});
