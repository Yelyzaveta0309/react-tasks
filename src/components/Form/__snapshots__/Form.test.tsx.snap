// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form render with snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-q0jhri-MuiInputBase-root-MuiInput-root"
    >
      <input
        class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        type="text"
        value=""
      />
    </div>
    <button
      disabled=""
      type="submit"
    >
      click
    </button>
  </form>
</DocumentFragment>
`;
